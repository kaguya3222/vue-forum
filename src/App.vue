<template>
  <div id="app">
    <nav-bar />
    <div class="container">
      <router-view />
    </div>
  </div>
</template>

<script>
import TheNavbar from "./components/TheNavbar";

export default {
  components: {
    "nav-bar": TheNavbar
  },
  beforeCreate() {
    this.$store.dispatch("users/fetchUser", {
      id: this.$store.state.users.authId
    });
  }
};
</script>

<style>
@import "./assets/css/style.css";
</style>
