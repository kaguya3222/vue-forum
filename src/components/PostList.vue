<template>
  <div class="post-list">
    <post-list-item v-for="post in posts" :key="post['.key']" :post="post" />
  </div>
</template>

<script>
import PostListItem from "./PostListItem";

export default {
  components: {
    "post-list-item": PostListItem
  },
  props: {
    posts: {
      required: true,
      type: Array
    }
  }
};
</script>

<style scoped></style>
