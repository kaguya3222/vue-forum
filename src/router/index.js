import Vue from "vue";
import VueRouter from "vue-router";

Vue.use(VueRouter);

const routes = [
  {
    path: "/",
    name: "Home",
    component: () => import("@/pages/PageHome")
  },
  {
    path: "/me",
    name: "Profile",
    component: () => import("@/pages/PageProfile")
  },
  {
    path: "/me/edit",
    name: "ProfileEditor",
    component: () => import("@/pages/PageProfile"),
    props: { edit: true }
  },
  {
    path: "/category/:id",
    name: "Category",
    component: () => import("@/pages/PageCategory"),
    props: true
  },
  {
    path: "/forum/:id",
    name: "Forum",
    component: () => import("@/pages/PageForum"),
    props: true
  },
  {
    path: "/thread/create/:forumId",
    name: "ThreadCreate",
    component: () => import("@/pages/PageThreadCreate"),
    props: true
  },
  {
    path: "/thread/:threadId",
    name: "ThreadShow",
    component: () => import("@/pages/PageThreadShow"),
    props: true
  },
  {
    path: "*",
    name: "NotFound",
    component: () => import("@/pages/PageNotFound")
  }
];

const router = new VueRouter({
  mode: "history",
  routes
});

export default router;
