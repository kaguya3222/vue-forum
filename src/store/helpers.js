import Vue from "vue";

const makeAppendChildToParentMutation = ({
  parents,
  child,
  parentModuleName
}) => (state, { childId, parentId, rootState }) => {
  const resource = rootState[parentModuleName][parents][parentId];
  if (!resource[child]) {
    Vue.set(resource, child, {});
  }
  Vue.set(resource[child], childId, childId);
};

export { makeAppendChildToParentMutation };
