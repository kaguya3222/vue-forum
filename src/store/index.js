import Vue from "vue";
import Vuex from "vuex";
import sourceData from "@/data.json";
import forumPosts from "./modules/posts/store";
import forumUsers from "./modules/users/store";

Vue.use(Vuex);

export default new Vuex.Store({
  state: {
    ...sourceData
  },
  getters: {},
  mutations: {},
  actions: {},
  modules: {
    forumPosts,
    forumUsers
  }
});
