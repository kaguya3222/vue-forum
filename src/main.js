import Vue from "vue";
import App from "./App.vue";
import "./registerServiceWorker";
import router from "./router";
import store from "./store";
import AppDate from "./components/AppDate";

Vue.config.productionTip = false;
Vue.component("app-date", AppDate);

new Vue({
  router,
  store,
  render: h => h(App)
}).$mount("#app");
