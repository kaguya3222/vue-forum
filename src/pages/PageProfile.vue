<template>
  <div class="flex-grid">
    <user-profile-card
      v-if="!edit"
      :user="user"
      :userPostsCount="userPostsCount"
      :userThreadsCount="userThreadsCount"
    />
    <user-profile-card-editor
      v-else
      :user="user"
      :userPostsCount="userPostsCount"
      :userThreadsCount="userThreadsCount"
    />
    <div class="col-7 push-top">
      <div class="profile-header">
        <span class="text-lead"> {{ user.username }}'s recent activity </span>
        <a href="#">See only started threads?</a>
      </div>

      <hr />
      <post-list :posts="userPosts" />
    </div>
  </div>
</template>

<script>
import PostList from "../components/PostList";
import { mapGetters } from "vuex";
import { countObjectProperties } from "../helpers";
import UserProfileCard from "../components/UserProfileCard";
import UserProfileCardEditor from "../components/UserProfileCardEditor";

export default {
  name: "PageProfile.vue",
  components: {
    "post-list": PostList,
    "user-profile-card": UserProfileCard,
    "user-profile-card-editor": UserProfileCardEditor
  },
  props: {
    edit: {
      default: false,
      type: Boolean
    }
  },
  computed: {
    ...mapGetters({
      user: "authUser",
      posts: "posts"
    }),
    userPosts() {
      if (this.user.posts) {
        return Object.values(this.posts).filter(
          post => post.userId === this.user[".key"]
        );
      } else {
        return [];
      }
    },
    userPostsCount() {
      return countObjectProperties(this.userPosts);
    },
    userThreadsCount() {
      return countObjectProperties(this.user.threads);
    }
  }
};
</script>

<style scoped></style>
