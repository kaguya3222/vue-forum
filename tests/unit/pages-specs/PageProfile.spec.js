import Vuex from "vuex";
import { createLocalVue, shallowMount } from "@vue/test-utils";
import PageProfile from "../../../src/pages/PageProfile";
import PostList from "../../../src/components/PostList";
import UserProfileCard from "../../../src/components/UserProfileCard";
import UserProfileCardEditor from "../../../src/components/UserProfileCardEditor";
import usersGetters from "../../../src/store/modules/users/getters";
import postsGetters from "@/store/modules/posts/getters";
import mockedSourceData from "../mocks/mockedSourceData";
import { countObjectProperties } from "../../../src/helpers";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("PageProfile", () => {
  const store = new Vuex.Store({
    state: { ...mockedSourceData, authId: "VXjpr2WHa8Ux4Bnggym8QFLdv5C3" },
    getters: {
      ...usersGetters,
      ...postsGetters
    }
  });
  const wrapper = shallowMount(PageProfile, {
    localVue,
    store
  });

  test("Shows user information", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("Accepts PostList as a child component", () => {
    expect(wrapper.contains(PostList)).toBe(true);
  });
  test("Accepts UserProfileCard as a child component", () => {
    expect(wrapper.contains(UserProfileCard)).toBe(true);
  });
  test("Accepts UserProfileCardEditor as a child component when 'edit' property is true", () => {
    wrapper.setProps({ edit: true });
    wrapper.vm.$nextTick(() => {
      expect(wrapper.contains(UserProfileCardEditor)).toBe(true);
    });
  });
  test("Accepts vuex authUser getter", () => {
    expect(wrapper.vm.user).toEqual(usersGetters.authUser(store.state));
  });
  test("Creates userPosts array of certain users", () => {
    const post = mockedSourceData.posts["-KsjnAAjzfpznUVh4DSa"];
    const hasPost = wrapper.vm.userPosts.some(
      userPost => userPost[".key"] === post[".key"]
    );
    expect(hasPost).toBe(true);
  });
  test("Correctly shows users posts number", () => {
    expect(wrapper.vm.userPostsCount).toBe(wrapper.vm.userPosts.length);
  });
  test("Correctly shows users threads number", () => {
    const userThreadsNumber = countObjectProperties(wrapper.vm.user.threads);
    expect(wrapper.vm.userThreadsCount).toBe(userThreadsNumber);
  });
});
