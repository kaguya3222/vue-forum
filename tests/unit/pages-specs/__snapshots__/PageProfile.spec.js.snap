// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageProfile Shows user information 1`] = `
<div class="flex-grid">
  <user-profile-card-stub user="[object Object]" userpostscount="9" userthreadscount="0"></user-profile-card-stub>
  <div class="col-7 push-top">
    <div class="profile-header"><span class="text-lead"> batman's recent activity </span> <a href="#">See only started threads?</a></div>
    <hr>
    <post-list-stub posts="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"></post-list-stub>
  </div>
</div>
`;
